_posts/2019-04-24-server-rendering-responsively.md

```diff
@@ -0,0 +1,199 @@
+
+One challenge we've run into with SSR is rendering markup and styles from the server that the browser doesn't have
+to throw away. When your browser finishes loading our JavaScript & hydrates our React app, our app suddenly has
+more information about your browser dimensions than we had in the server request. React will go through the
+reconciliation process with this new information, to determine which DOM elements need to be updated. If there are
+significant differences between the markup our server gave you and the component tree our React app is now
+rendering, there's noticeable churn in the DOM.
+
+We'd like our app to churn as little as possible during hydration. We've taken several steps (and mis-steps) toward
+accomplishing this goal: how can we emit _usable_ responsive markup from the server, so that the browser doesn't
+have to throw everything away when the browser dimensions become known?
+
+To help us solve this problem, we've released
+[@artsy/react-responsive-media](https://github.com/artsy/react-responsive-media). This is the story of how
+@artsy/react-responsive-media came about.
+
+<!-- more -->
+
+## Iteration 1: Render props and a disposable component tree
+
+Our initial attempt to emit responsive content from the server is what prompted us to dig deeper into this problem.
+We started with a `<Responsive>` component that used
+[a "render props"/"function as a child" pattern](https://reactjs.org/docs/render-props.html#using-props-other-than-render),
+and allowed you to return different JSX based on the current breakpoint. In practice, it looked something like
+this:
+
+```
+<Responsive>
+  {({ xs }) => {
+    if (xs) {
+      return <Box width={1} />
+    } else {
+      return <Box width={2} />
+    }
+  }}
+</Responsive>
+```
+
+The argument to the child function contained a flag for each of the breakpoints our app supports. In the example
+above, we're rendering a `<Box>` with a width of `1` if the screen matches our `xs` breakpoint; otherwise, we're
+rendering a `<Box>` with a width of `2`.
+
+Our biggest problem with this API was that on the server, we don't know which breakpoint you're at. We had to
+default to something, so we assumed that if we didn't know which breakpoint you wanted, we would give you the
+`xs`/"extra small" one. But if you were browsing with a desktop, the first client-side render will completely throw
+away what the server gave you. In the example above, that's probably not that big a deal, but our `<Responsive>`
+elements tended to contain large sub-trees of components. This is a big deal, and this is the DOM churn we really
+want to avoid.
+
+In addition, this API was imperative. Sprinkling our app with lots of `if/else/else` statements degraded
+readability, and it just generally felt against the spirit of building a declarative React app. We also avoid the
+render props pattern when possible because we find it less readable.
```

Comment:
It would be good to mention prior art in this paragraph. There's [react-media](https://github.com/ReactTraining/react-media) and https://github.com/contra/react-responsive. Both of these libs are similar in design to our V1 component.